src/ray/raylet/node_manager.cc

```diff
@@ -3000,76 +3001,120 @@ void NodeManager::FinishAssignedActorCreationTask(const ActorID &parent_actor_id
-              << " not found in either GCS or lineage cache. It may have been evicted "
-              << "by the redis LRU configuration. Consider increasing the memory "
-                 "allocation via "
-              << "ray.init(redis_max_memory=<max_memory_bytes>).";
-          MarkObjectsAsFailed(ErrorType::OBJECT_UNRECONSTRUCTABLE, {required_object_id},
-                              JobID::Nil());
-        }
-      }));
+  // Get the owner's address.
+  rpc::Address owner_addr;
+  bool still_required =
+      task_dependency_manager_.GetOwnerAddress(required_object_id, &owner_addr);
+  if (!still_required) {
+    // The object is no longer required on this node. If another
+    // client needs the object later on, the timeout will be reset.
+    return;
+  }
+
+  if (!owner_addr.worker_id().empty()) {
+    // The owner's address exists. Poll the owner to check if the object is
+    // still in scope. If not, mark the object as failed.
+    // TODO(swang): If the owner has died, we could also mark the object as
+    // failed as soon as we hear about the owner's failure from the GCS,
+    // avoiding the raylet's reconstruction timeout.
+    auto client = std::unique_ptr<rpc::CoreWorkerClient>(
+        new rpc::CoreWorkerClient(owner_addr, client_call_manager_));
+
+    rpc::GetObjectStatusRequest request;
+    request.set_object_id(required_object_id.Binary());
+    request.set_owner_worker_id(owner_addr.worker_id());
+    RAY_CHECK_OK(client->GetObjectStatus(
+        request, [this, required_object_id](Status status,
+                                            const rpc::GetObjectStatusReply &reply) {
+          if (!status.ok() || reply.status() == rpc::GetObjectStatusReply::OUT_OF_SCOPE) {
+            // The owner is gone or the owner replied that the object has gone
+            // out of scope (this is an edge case in the distributed ref counting
+            // protocol where a borrower dies before it can notify the owner of
+            // another borrower). Store an error in the local plasma store so
+            // that an exception will be thrown when the worker tries to get the
+            // value.
+            MarkObjectsAsFailed(ErrorType::OBJECT_UNRECONSTRUCTABLE, {required_object_id},
+                                JobID::Nil());
+          }
+          // Do nothing if the owner replied that the object is available. The
+          // object manager will continue trying to fetch the object, and this
+          // handler will get triggered again if the object is still
+          // unavailable after another timeout.
+        }));
+  } else {
+    // We do not have the owner's address. This is either an actor creation
+    // task or a randomly generated ObjectID. Try to look up the spec for the
+    // actor creation task.
+    // TODO(swang): The task lookup is only needed when the GCS actor service is
+    // disabled. Once the GCS actor service is enabled by default, we can
```

Comment:
I'm wondering if we should just not implement this given that GCS actor service is coming soon, and this is flagged off by default?

Or do we rely on this for the normal actor reconstruction path?

Reply:
It's to make sure that actor reconstruction works when the GCS actor service is flagged off. We can remove it once the GCS actor service is the default.